crates/primitives/src/transaction/tx_type.rs

```diff
@@ -54,25 +64,38 @@ impl From<TxType> for U8 {
 }
 
 impl Compact for TxType {
-    fn to_compact<B>(self, _: &mut B) -> usize
+    fn to_compact<B>(self, buf: &mut B) -> usize
     where
         B: bytes::BufMut + AsMut<[u8]>,
     {
         match self {
             TxType::Legacy => 0,
             TxType::EIP2930 => 1,
             TxType::EIP1559 => 2,
-            TxType::EIP4844 => 3,
+            _ => {
+                buf.put_u8(self as u8);
+                3
+            }
         }
     }
 
-    fn from_compact(buf: &[u8], identifier: usize) -> (Self, &[u8]) {
+    // For backwards compatibility purposes only 2 bits of the type are encoded in the identifier
+    // parameter. In the case of a 3, the full transaction type is read from the buffer as a
+    // single byte.
+    fn from_compact(mut buf: &[u8], identifier: usize) -> (Self, &[u8]) {
         (
             match identifier {
                 0 => TxType::Legacy,
                 1 => TxType::EIP2930,
                 2 => TxType::EIP1559,
-                _ => TxType::EIP4844,
+                _ => {
+                    let identifier = buf.get_u8() as usize;
+                    match identifier {
+                        #[cfg(feature = "optimism")]
+                        126 => TxType::DEPOSIT,
+                        _ => TxType::EIP4844,
+                    }
+                }
```

Comment:
same

Reply:
@mattsse how can we make this usize match exhaustive?